src/Functions/MatchImpl.h

```diff
@@ -391,22 +394,248 @@ struct MatchImpl
 
                 /// Tail, in which there can be no substring.
                 if (i < res.size())
-                    memset(&res[i], revert, (res.size() - i) * sizeof(res[0]));
+                    memset(&res[i], negate, (res.size() - i) * sizeof(res[0]));
             }
         }
     }
 
-    template <typename... Args>
-    static void vectorVector(Args &&...)
+    static void vectorVector(
+        const ColumnString::Chars & haystack_data,
+        const ColumnString::Offsets & haystack_offsets,
+        const ColumnString::Chars & needle_data,
+        const ColumnString::Offsets & needle_offset,
+        const ColumnPtr & start_pos_,
+        PaddedPODArray<UInt8> & res)
     {
-        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Function '{}' doesn't support non-constant needle argument", name);
+        const size_t haystack_size = haystack_offsets.size();
+
+        if (haystack_size != needle_offset.size())
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Function '{}' unexpectedly received a different number of haystacks and needles", name);
+
+        if (start_pos_ != nullptr)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Function '{}' doesn't support start_pos argument", name);
+
+        if (haystack_offsets.empty())
+            return;
+
+        String required_substr;
+        bool is_trivial;
+        bool required_substring_is_prefix; /// for `anchored` execution of the regexp.
+
+        size_t prev_haystack_offset = 0;
+        size_t prev_needle_offset = 0;
+
+        for (size_t i = 0; i < haystack_size; ++i)
+        {
+            const auto * const cur_haystack_data = &haystack_data[prev_haystack_offset];
+            const size_t cur_haystack_length = haystack_offsets[i] - prev_haystack_offset - 1;
+
+            const auto * const cur_needle_data = &needle_data[prev_needle_offset];
+            const size_t cur_needle_length = needle_offset[i] - prev_needle_offset - 1;
+
+            const auto & needle = String(
+                    reinterpret_cast<const char *>(cur_needle_data),
+                    cur_needle_length);
+
+            if (like && impl::likePatternIsSubstring(needle, required_substr))
+            {
+                if (required_substr.size() > cur_haystack_length)
+                    res[i] = negate;
+                else
+                {
+                    Searcher searcher(required_substr.data(), required_substr.size(), cur_haystack_length);
+                    const auto * match = searcher.search(cur_haystack_data, cur_haystack_length);
+                    res[i] = negate
+                        ^ (match != cur_haystack_data + cur_haystack_length);
+                }
+            }
+            else
+            {
+                // each row is expected to contain a different like/re2 pattern
+                // --> bypass the regexp cache, instead construct the pattern on-the-fly
```

Comment:
* Could it be good to check if cache maybe already contain a pattern anyway? If not, we can compile, but do not push a value into the cache.
* Maybe we can create local cache for query/function? We will need eviction for it, but it may be useful if there are a few patterns (like, in case `select ... where something like (flag ? '%pattern1%' : '%pattern2%'`)
* Probably we can update global cache anyway, got the limited number of patterns. Or we can use another one global cache especially for non-constant needle. Or we can support eviction for single global cache.

Reply:
If we want to implement something of it, it would be better to do in a separate PR

Reply:
These optimizations surely make sense in some form or the other. As you said, I would defer that to a separate PR too.

Reply:
Yes, the case of a few alternative regexps makes sense.